### DISCUSSION TITLE
Issues with ViewLocator Navigating Views after Publishing

### DISCUSSION THREAD

**Discussion by ihatebananas** [AvaloniaUI/Avalonia#16438] [Q&A]
I have a simple application that essentially accepts directories from the user, after which the user continues onto a new page to perform error checks. In order to navigate views, I am currently using the ViewLocator, and it works perfectly well when I run the application from within Visual Studio. 

For reference, this is the relevant portion in MainWindowViewModel.cs:

```
public class MainWindowViewModel : ViewModelBase
{
    public ReactiveCommand<string[], Unit> NextCommand { get; }
    private ViewModelBase _contentViewModel;

    public MainWindowViewModel()
    {
        _contentViewModel = new LoadDataViewModel();
        NextCommand = ReactiveCommand.Create<string[]>(NextScreen);
    }

    public ViewModelBase ContentViewModel
    {
        get => _contentViewModel;
        private set => this.RaiseAndSetIfChanged(ref _contentViewModel, value);
    }

    public void NextScreen(string[] directories)
    {
        ContentViewModel = new RunCheckViewModel(directories[0], directories[1]);
    }
}
```
Here, ContentViewModel is bound to the "Content" property of the Window tag in MainWindow.axaml. 

I have a "Continue" button that is bound to the NextCommand command. When run from within Visual Studio, pressing "Continue" changes the view to the "RunCheckView", as one would expect from the ViewLocator. However, after publishing the application, pressing "Continue" does nothing. The application does not crash, and the UI, outside of the "Continue" button, works perfectly fine. It just doesn't navigate to the RunCheckView. As I said earlier, this does not happen when run from within Visual Studio.

I have tried searching this up and saw some people attributed this to trimming, and so I tried the following:
1. Set TrimMode to "copyused" in the .csproj file.
2. Set TrimMode to "partial" in the .csproj file.
3. Set PublishTrimmed to "false" in the .csproj file. 

I saw the first entry being used in the Avalonia sample for navigating views, but it didn't work. The other two didn't either. As for how I published it, I just did `dotnet publish MyProject.csproj -r win-x64`.

To be clear, I am absolutely sure that the command for switching views is working properly -- the "NextScreen" method in MainWindowViewModel is being called properly (it properly prints debug print statements I inserted). However, changing the ContentViewModel does not change the view itself. Again, this only happens after publishing; this is not an issue when run from within Visual Studio. 

If it matters, I am on Windows, but this is set up as a cross-platform application. I'm using NET 8.0 as well. 

So, what can I do about this?

Thanks in advance, any help is appreciated.

**Comment by thevortexcloud** [comment]
Have you double checked that you don't have AOT enabled?

**Reply by ihatebananas** [reply]
I wasn't aware of such a thing, so I didn't explicitly do anything about it. Since seeing your suggestion, I've tried adding the following to the .csproj file for the application:
1. `<RunAOTCompilation>false</RunAOTCompilation>`
2. `<PublishAot>false</PublishAot>`

I'm fairly new to Avalonia and app creation in general, so I'm not sure if this is what you had intended. Either way, the tags I listed above did not change the behavior of the Continue button. I also tried including `<PublishTrimmed>false</PublishTrimmed>`, to no avail. Am I doing this wrong, or is something else the issue?

**Reply by ihatebananas** [reply]
Sorry, this was a non-issue. It seems the call to initialize RunCheckViewModel threw an exception because of incorrect file I/O. I was using relative filepaths using the bin/Debug/net8.0 folder, which works in Visual Studio, but of course the working directory is different once published. I manually added the files I needed to the folder after publishing, and it works just fine.

**Reply by thevortexcloud** [reply]
For future reference: If you are running on windows, .NET by default dumps unhandled exceptions to the Windows event application log.
